drivers/sdhc/sdhc_stm32.c

```diff
@@ -0,0 +1,797 @@
+							dev_data->total_transfer_bytes);
+			res = HAL_SDIO_WriteExtended_DMA(config->hsd, &arg,
+							dev_data->sdio_dma_buf,
+							dev_data->total_transfer_bytes);
+		}
+	} else {
+		if (IS_ENABLED(CONFIG_SDHC_STM32_POLLING_SUPPORT)) {
+			res = HAL_SDIO_ReadExtended(config->hsd, &arg, data->data,
+							dev_data->total_transfer_bytes,
+							data->timeout_ms);
+		} else {
+			sys_cache_data_flush_range(dev_data->sdio_dma_buf,
+						dev_data->total_transfer_bytes);
+			res = HAL_SDIO_ReadExtended_DMA(config->hsd, &arg, dev_data->sdio_dma_buf,
+							dev_data->total_transfer_bytes);
+		}
+	}
+
+	if (!IS_ENABLED(CONFIG_SDHC_STM32_POLLING_SUPPORT)) {
+		/* Wait for whole transfer to complete */
+		if (k_sem_take(&dev_data->device_sync_sem, K_MSEC(CONFIG_SD_CMD_TIMEOUT)) != 0) {
+			k_free(dev_data->sdio_dma_buf);
+			return -ETIMEDOUT;
+		}
+
+		if (direction == SDIO_IO_READ) {
+			sys_cache_data_invd_range(dev_data->sdio_dma_buf,
+						dev_data->total_transfer_bytes);
+			memcpy(data->data, dev_data->sdio_dma_buf, data->block_size * data->blocks);
+		}
+
+		k_free(dev_data->sdio_dma_buf);
+	}
+
+	return res == HAL_OK ? 0 : -EIO;
+}
+
+static int sdhc_stm32_switch_to_1_8v(const struct device *dev)
+{
+	uint32_t res;
+	struct sdhc_stm32_data *data = dev->data;
+	const struct sdhc_stm32_config *config = dev->config;
+
+	/* Check if host supports 1.8V signaling */
+	if (!data->props.host_caps.vol_180_support) {
+		LOG_ERR("Host does not support 1.8V signaling");
+		return -ENOTSUP;
+	}
+
+	res = SDMMC_CmdVoltageSwitch(config->hsd->Instance);
+	if (res != 0) {
+		LOG_ERR("CMD11 failed: %#x", res);
+		return -EIO;
+	}
+
+	LOG_DBG("Successfully switched to 1.8V signaling");
+	return 0;
+}
+
+static int sdhc_stm32_request(const struct device *dev, struct sdhc_command *cmd,
+			      struct sdhc_data *data)
+{
+	int res = 0;
+	uint32_t sdmmc_res = 0U;
+	struct sdhc_stm32_data *dev_data = dev->data;
+	const struct sdhc_stm32_config *config = dev->config;
+
+	__ASSERT_NO_MSG(cmd != NULL);
+
+	if (k_mutex_lock(&dev_data->bus_mutex, K_MSEC(cmd->timeout_ms))) {
+		return -EBUSY;
+	}
+
+	if (HAL_SDIO_GetState(config->hsd) != HAL_SDIO_STATE_READY) {
+		LOG_ERR("SDIO Card is busy");
+		k_mutex_unlock(&dev_data->bus_mutex);
+		return -ETIMEDOUT;
+	}
+
+	(void)pm_device_runtime_get(dev);
+
+	/* Prevent the clocks to be stopped during the request */
+	pm_policy_state_lock_get(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+
+	dev_data->cmd_index = cmd->opcode;
+	switch (cmd->opcode) {
+	case SD_GO_IDLE_STATE:
+		sdmmc_res = sdhc_stm32_go_idle_state(dev);
+		if (sdmmc_res != 0U) {
+			res = -EIO;
+		}
+		break;
+
+	case SD_SELECT_CARD:
+		sdmmc_res = SDMMC_CmdSelDesel(config->hsd->Instance, cmd->arg);
+		if (sdmmc_res == 0U) {
+			/* Clear unused flags to avoid SDIO card identification issues */
+			cmd->response[0U] &=
+				~(SD_R1_ERASE_SKIP | SD_R1_CSD_OVERWRITE | SD_R1_ERASE_PARAM);
+		} else {
+			res = -EIO;
+		}
```

Comment:
swap the condition, if it fails set `res` and `break`, otherwise execute inline, same below